sdk/core/azure-core/inc/azure/core/internal/tracing/service_tracing.hpp

```diff
@@ -193,27 +193,48 @@ namespace Azure { namespace Core { namespace Tracing { namespace _internal {
     {
     }
 
-    DiagnosticTracingFactory() = default;
-    DiagnosticTracingFactory(DiagnosticTracingFactory const&) = default;
+    ContextAndSpanFactory() = default;
+    ContextAndSpanFactory(ContextAndSpanFactory const&) = default;
 
     /** @brief A ContextAndSpan provides an updated Context object and a new span object
      * which can be used to add events and attributes to the span.
      */
     using ContextAndSpan = std::pair<Azure::Core::Context, ServiceSpan>;
 
+    /**
+     * @brief Create a span with the specified span name.
+     *
+     * @details This method is a convenience method intended for use by service clients, it creates
+     * a SpanKind::Internal span and context.
+     *
+     * @param spanName Name for the span to be created.
+     * @param clientContext parent context object.
+     *
+     * @returns Newly allocated context and Span object.
+     *
+     */
     ContextAndSpan CreateSpan(
         std::string const& spanName,
-        Azure::Core::Tracing::_internal::SpanKind const& spanKind,
-        Azure::Core::Context const& clientContext);
+        Azure::Core::Context const& clientContext) const;
 
+    /**
+     * @brief Create a span with the specified span name and create options.
+     *
+     * @param spanName Name for the span to be created.
+     * @param spanOptions Options for the newly created span.
+     * @param clientContext parent context object.
+     *
+     * @returns Newly allocated context and Span object.
+     *
+     */
     ContextAndSpan CreateSpan(
         std::string const& spanName,
         Azure::Core::Tracing::_internal::CreateSpanOptions& spanOptions,
-        Azure::Core::Context const& clientContext);
+        Azure::Core::Context const& clientContext) const;
 
-    std::unique_ptr<Azure::Core::Tracing::_internal::AttributeSet> CreateAttributeSet();
+    std::unique_ptr<Azure::Core::Tracing::_internal::AttributeSet> CreateAttributeSet() const;
 
-    static std::unique_ptr<DiagnosticTracingFactory> DiagnosticFactoryFromContext(
+    static std::unique_ptr<ContextAndSpanFactory> ContextAndSpanFactoryFromContext(
```

Comment:
I am glad this class is internal. That said, a method name like this makes me think the factory class name might need another look.

Reply:
Hmm, so context and span is literally an alias for a pair of context and span. I see.

Reply:
Could you explain further? The ContextAndSpanFactory creates ContextAndSpan objects. This method creates a ContextAndSpanFactory from a Context object (we have situations where we need to be able to retrieve the ContextAndSpanFactory for the current client from a pipeline policy and the Context object is the only thing that ties the client and the policies together).

Reply:
We discussed and brainstormed some ideas offline.

**Conclusion:**
- Name `DiagnosticTracingFactory`​ as: ​`TracingContextFactory`​.
- Rename the static method ​`DiagnosticFactoryFromContext`​ to ​`CreateFromContext​`.
- Change ​`std::pair`​ ​`ContextAndSpan`​ into its own more extensible/future-proof ​`TracingContext​` type.
- Rename factory methods from `​CreateSpan​` to ​`CreateTracingContext​`.
- It is fine to have an internal returning type method on ​`TracingProvider​`, it is the exception to the rule. We can't hide `CreateTracer` from the user and no C++ keyword is sufficient here (private or protected virtual).
- For later: Some doc comment improvement suggestions (note about the ​`CreateTracer​` method intended for internal use, file comment at the top of `​tracing.hpp`​, why we delete copy ctors on ​`ServiceSpan`​​, etc.).
- Leave the header structure as-is (no need to split-off tracing.hpp internal types into its own header, used by an open telemetry provider implementer).

Reply:
We'd want to update this alias to no longer be called `TracingContext` to avoid confusion with the Context and Span type now named `TracingContext`:
https://github.com/Azure/azure-sdk-for-cpp/blob/48ef687a8f941a6ab755b1d4e56701edcb8e29cf/sdk/core/azure-core/inc/azure/core/internal/tracing/service_tracing.hpp#L180-L181